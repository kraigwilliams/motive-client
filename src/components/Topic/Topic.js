/* eslint-disable eqeqeq */
import React, { Component } from "react";
import ContentContext from "./../../contexts/ContentContext";
import ContentService from "../../services/content-service";
import TokenService from "../../services/token-service";
import {
  TopicHeader,
  TopicWrapper,
  ContentWrapper,
  SortWrapper,
  SortSelectDropdown,
  SortLabel,
  StyledDeleteDiv,
} from "./Topic.style";
import CondensedThought from "../CondensedThought/CondensedThought";
import { AddButton, GoBack } from "../Button/Button";
import {
  DeleteButton,
  ConfirmDeleteButton,
  ShareButton,
} from "../Button/Button";
import { colors } from "../constants";

export default class Topic extends Component {
  static contextType = ContentContext;

  constructor(props) {
    super(props);
    this.state = {
      currentTopic: {},
      currentThoughts: [],
      deleteDiv: true,
      topicId: null,
      isShared: null,
      sharedLevel: null,
      sort_option: 1,
    };
  }

  async componentDidMount() {
    //get the topic id and auth token of the user
    const topicId = await this.props.match.params.topic_id;
    this.setState({
      topicId,
    });
    // localStorage.setItem('topic', topicId);
    this.context.setTopicForAddThought(topicId);
    this.context.setTopicForThought(topicId);
    const authToken = TokenService.getAuthToken();

    //get the current topic from the server and set it in state
    const currentTopic = await ContentService.getThisTopic(topicId, authToken);
    this.setState({ currentTopic });

    //get the thoughts that belong to the current topic and set it in state
    const currentThoughts = await ContentService.getThoughtsInTopic(
      topicId,
      this.state.sort_option
    );
    if (currentThoughts) {
      this.setState({ currentThoughts });
    }

    const sharedTopics = await ContentService.getSharedTopics();

    const isShared = sharedTopics.find((topic) => topic.topic_id == topicId);

    if (isShared !== undefined) {
      const level = await ContentService.getSharedTopicLevel(topicId);

      this.setState({
        isShared: true,
        sharedLevel: level,
      });
    } else {
      this.setState({
        isShared: false,
      });
    }
  }

  toggleDeleteDiv = () => {
    this.setState({
      deleteDiv: !this.state.deleteDiv,
    });
  };

  handleDelete = () => {
    ContentService.deleteTopic(this.state.topicId);
    this.props.history.goBack();
  };

  handleGoBack = (ev) => {
    this.context.setTopicForAddThought(null);
    this.context.setTopicForThought(0);
    this.props.history.goBack();
  };

  async handleSortChange(ev) {
    ev.preventDefault();
    const sort_option = ev.target.value;

    this.setState({
      sort_option,
    });

    const currentThoughts = await ContentService.getThoughtsInTopic(
      this.state.topicId,
      sort_option
    );
    if (currentThoughts) {
      this.setState({ currentThoughts });
    }
    console.log(currentThoughts, "current thoughts in sort change handler");
  }

  render() {
    const { currentTopic, currentThoughts, topicId, sharedLevel } = this.state;

    return (
      <TopicWrapper>
        <div style={{ display: "flex", overflow: "hidden" }}>
          <GoBack
            type="reset"
            onClick={() => this.handleGoBack()}
            margin="30px 0px 30px 30px"
          />
          <TopicHeader>
            <h1 style={{ marginBottom: "20px" }}>
              {currentTopic.topic_title}!
            </h1>
            <p style={{ color: colors.darkgrey }}>
              {currentTopic.topic_content}
            </p>
          </TopicHeader>
          <div
            style={{
              display: "flex",
              flexDirection: "column",
              marginTop: "40px",
            }}
          >
            <ShareButton
              color={colors.blue}
              type="button"
              to={`/topic/${topicId}/share`}
              disabled={sharedLevel > 2}
              data-tip="Share Topic"
            />

            {currentThoughts.length < 1 ? (
              <DeleteButton
                color={colors.blue}
                data-tip="Delete Topic"
                type="button"
                onClick={this.toggleDeleteDiv.bind(this)}
              />
            ) : (
              ""
            )}
          </div>
        </div>

        <ContentWrapper>
          <div className="top">
            <SortWrapper>
              <SortLabel htmlFor="sort-dropdown">Sort</SortLabel>
              <SortSelectDropdown
                onChange={this.handleSortChange.bind(this)}
                name="sortOptions"
                id="sort-dropdown"
                aria-label="sort alphabetically or by date created by choosing from this dropdown"
              >
                <option value="1">Most Recent</option>
                <option value="2">A-Z</option>
              </SortSelectDropdown>
            </SortWrapper>

            <AddButton
              type="button"
              to="/add-thought"
              data-tip="Add Thought to Topic"
            />
          </div>
          <div className="topic-thoughts">
            {currentThoughts.map((thought, idx) => {
              return (
                <CondensedThought
                  key={idx}
                  id={thought.id}
                  title={thought.thought_title}
                  topic={topicId}
                />
              );
            })}
          </div>

          {!this.state.deleteDiv && (
            <StyledDeleteDiv>
              {" "}
              Delete Topic?
              <ConfirmDeleteButton type="button" onClick={this.handleDelete}>
                Yes
              </ConfirmDeleteButton>
            </StyledDeleteDiv>
          )}
        </ContentWrapper>
      </TopicWrapper>
    );
  }
}
