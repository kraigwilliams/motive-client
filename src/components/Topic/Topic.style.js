import styled from "styled-components";
import { colors } from "../constants";

export const TopicWrapper = styled.div`
  display: flex;
  flex-direction: column;
  background-color: ${colors.slategrey};
  margin: 0;
  padding: 0;
  min-height: 100vh;
  .header-btn {
    width: 28px;
    justify-content: space-between;
  }
`;

export const TopicHeader = styled.header`
  text-align: center;
  padding: 30px 30px 10px 30px;
  color: ${colors.white};
  text-align: center;
  margin: 0;
  width: 100%;
  justify-self: center;
  display: flex;
  flex-direction: column;
  overflow: hidden;

  @media (min-width: 600px) {
    text-align: center;
  }
`;

export const SortWrapper = styled.div`
  display: flex;
  flex-direction: column;
  text-align: center;
  width: fit-content;
`;

export const SortLabel = styled.label`
  padding-bottom: 5px;
  color: ${colors.coral};
  margin-bottom: 5px;
  margin-left: 70%;
  margin: auto;
  font-size: 16px;
  font-weight: bolder;

  @media (min-width: 600px) {
    margin-right: auto;
    margin-bottom: 10px;
  }
`;

export const SortSelectDropdown = styled.select`
  background: transparent;
  font-size: 16px;
  color: ${colors.white};

  > option {
    font-size: 16px;
  }

  @media (min-width: 600px) {
    min-width: 30%;
    margin-left: auto;
    margin-right: auto;
  }
`;

export const ContentWrapper = styled.div`
  display: flex;
  flex-direction: column;
  background-color: ${colors.darkgrey};
  padding: 30px;
  margin: 0;
  height: 100vh;

  .top {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    /* margin: 10px 30px; */
    margin: 10px 0px;
    /* padding: 5px 60px; */
    padding: 5px 0px;
  }

  .topic-thoughts {
    width: 80%;
    margin: 0px auto;
  }

  @media (min-width: 600px) {
    .topic-thoughts {
      width: 70%;
    }
  }
`;

export const StyledDeleteDiv = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  text-align: center;
  border-radius: 5px;
  padding: 25px;
  background-color: ${colors.darkgrey};
  color: ${colors.offwhite};
  /* text-align: left; */

  @media (min-width: 600px) {
  }
`;
