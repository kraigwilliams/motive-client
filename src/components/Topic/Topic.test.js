import React from 'react'
import { shallow, configure } from 'enzyme'
import toJson from 'enzyme-to-json'
import Topic from './Topic'
import Adapter from 'enzyme-adapter-react-16';


configure({adapter: new Adapter()});
describe(`renders <Topic />`, () => {
  //Snapshot Testing
  it('renders Topic by default', () => {
    const wrapper = shallow(
      <Topic />
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})