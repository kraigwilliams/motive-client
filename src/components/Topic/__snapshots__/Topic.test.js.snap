// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Topic /> renders Topic by default 1`] = `
<styled.div>
  <div
    style={
      Object {
        "display": "flex",
        "overflow": "hidden",
      }
    }
  >
    <Styled(Component)
      margin="30px 0px 22px 30px"
      onClick={[Function]}
      type="reset"
    />
    <styled.div>
      <h1
        style={
          Object {
            "marginBottom": "20px",
          }
        }
      />
      <p
        style={
          Object {
            "color": "#424B54",
          }
        }
      />
    </styled.div>
    <div
      style={
        Object {
          "width": "66.97px",
        }
      }
    />
  </div>
  <styled.div>
    <div
      className="top"
    >
      <styled.div>
        <styled.h3>
          Sort
        </styled.h3>
        <styled.select>
          <option
            value="A-Z"
          >
            A-Z
          </option>
          <option
            value="Most Recent"
          >
            Most Recent
          </option>
        </styled.select>
      </styled.div>
      <Styled(Component)
        to="/add-thought"
        type="button"
      />
    </div>
    <Styled(Component)
      onClick={[Function]}
      type="button"
    />
  </styled.div>
</styled.div>
`;

exports[`renders <Topic /> renders Topic by default 1`] = `
<styled.div>
  <div
    style={
      Object {
        "display": "flex",
        "overflow": "hidden",
      }
    }
  >
    <Styled(Component)
      margin="30px 0px 30px 30px"
      onClick={[Function]}
      type="reset"
    />
    <styled.div>
      <h1
        style={
          Object {
            "marginBottom": "20px",
          }
        }
      />
      <p
        style={
          Object {
            "color": "#424B54",
          }
        }
      />
    </styled.div>
    <Styled(Component)
      disabled={false}
      to="/topic/null/share"
      type="button"
    />
  </div>
  <Styled(Component)
    onClick={[Function]}
    type="button"
  />
  <styled.div>
    <div
      className="top"
    >
      <styled.div>
        <styled.h3>
          Sort
        </styled.h3>
        <styled.select
          name="sortOptions"
          onChange={[Function]}
        >
          <option
            value="1"
          >
            Most Recent
          </option>
          <option
            value="2"
          >
            A-Z
          </option>
        </styled.select>
      </styled.div>
      <Styled(Component)
        to="/add-thought"
        type="button"
      />
    </div>
  </styled.div>
</styled.div>
`;
